2.1 what's new (#4645)
sunpy/sunpy: docs/whatsnew/2.1.rst

Alongside this, there is easier access to the results from a client by indexing the results by the client name::

    >>> hek_results, jsoc_results = results['hek'], results['jsoc']

In the first example you can see the search results from HEK are not too useful by default.
The results are truncated as the long urls for the thumbnail take up too many characters and HEK can return up to 100 columns of information.
With a new method ``show()`` on the results itself, you directly select the columns to display::

    >>> hek_results.show('event_peaktime', 'obs_instrument', 'fl_peakflux')
    event_peaktime   obs_instrument fl_peakflux
    ------------------- -------------- -----------
    2010-08-01T03:40:37            AIA     1027.64
    2010-08-01T03:40:44            AIA     1441.78

To find the keys for all of the columns you can do::

    >>> hek_results.keys()
    ['gs_thumburl',
    'comment_count',
    'hpc_bbox',
    ...
    'area_unit',
    'obs_lastprocessingdate',
    'refs']

In addition, if you pass the entire search query into ``Fido.fetch()`` and it will ignore results that have no corresponding data files to retrieve.

.. _whatsnew-2.1-refactor:

> Fido Results Refactor

Reviewer: I can understand not wanting to make all of the mentions of the Fido methods be references, e.g.:
```
:meth:`sunpy.net.Fido.search`
```
but it's a bit odd that none of them are.

=====================

While working on the above change, several changes were made to the way that results are returned from ``Fido.search()`` and the search methods of the underlying clients.

We have tried to minimize any breaking changes here and we believe that for most users the difference between 2.0 and 2.1 will be minor.

The key highlights you will want to be aware of are:

* Previously slicing the result of ``Fido.search()`` (a `~sunpy.net.fido_factory.UnifiedResponse` object) so that it had a length of one returned another `~sunpy.net.fido_factory.UnifiedResponse` object, now it will return a `~sunpy.net.base_client.QueryResponseTable` object, which is a subclass of `astropy.table.Table`.

* All result objects contained within the results of a ``Fido.search()`` are now `~sunpy.net.base_client.QueryResponseTable` objects (or subclasses thereof).

>   These objects are subclasses of `astropy.table.Table` and can therefore be filtered and inspected as tabular objects, and the modified tables can be passed to ``Fido.fetch``.

Reviewer: ```suggestion
  These objects are subclasses of `astropy.table.Table` and can therefore be filtered and inspected as tabular objects, and the modified tables can be passed to ``Fido.fetch()``.
```

* The keys available to be used when formatting the ``path=`` argument to ``Fido.fetch()`` have changed. This is to standardise them over the results from more clients and make them easier to use.
  You can use the `~.UnifiedResponse.path_format_keys` method to see all the possible keys for a particular search.

> * The search results object returned from ``Fido.search`` now correctly counts all results in it's `~sunpy.net.fido_factory.UnifiedResponse.file_num` property.

Reviewer: ```suggestion
* The search results object returned from ``Fido.search`` now correctly counts all results in its `~sunpy.net.fido_factory.UnifiedResponse.file_num` property.
```

> * Results from the `~sunpy.net.dataretriever.NOAAIndicesClient` and the `~sunpy.net.dataretriever.NOAAPredictClient` no longer have ``Start Time`` or ``End Time`` in their results table as the results returned from the client are not dependant upon the time parameter of a search.

Reviewer: ```suggestion
* Results from the `~sunpy.net.dataretriever.NOAAIndicesClient` and the `~sunpy.net.dataretriever.NOAAPredictClient` no longer have ``Start Time`` or ``End Time`` in their results table as the results returned from the client are not dependent upon the time parameter of a search.
```

.. _whatsnew-2.1-contours:

> Finding map contours

Reviewer: I think this entry should be pushed down to between the reprojection entry and the performance entry

====================

> The new :meth:`sunpy.map.contour` method can be used to extract contours from a map. It returns contours as a `~astropy.coordinates.SkyCoord`, allowing contours to be easily overplotted on the original or other maps.

Reviewer: ```suggestion
The new :meth:`sunpy.map.GenericMap.contour` method can be used to extract contours from a map. It returns contours as a `~astropy.coordinates.SkyCoord`, allowing contours to be easily overplotted on the original or other maps.
```

.. _whatsnew-2.1-unitdata:

> Map data and units

Reviewer: This seems like a rather unremarkable entry for a What's New...

==================

> `~sunpy.map.Map`s now have the ``.unit`` set to the correct `~astropy.units.Unit` if the 'BUNIT' keyword is set in the FITS metadata.

Reviewer: ```suggestion
`~sunpy.map.Map` instances now have the ``.unit`` set to the correct `~astropy.units.Unit` if the 'BUNIT' keyword is set in the FITS metadata.
```
It looks like the backtick followed by an "s" is misinterpreted by Sphinx

.. _whatsnew-2.1-reprojecting:

Alternatives for reprojecting a Helioprojective map
===================================================
The typical observation in `~sunpy.coordinates.frames.Helioprojective` coordinates does not contain full 3D information for the sources of emission, so an assumption needs to be made when transforming such coordinates to other coordinate frames.
By default, SunPy assumes that the emission is coming from the surface of the Sun, which enables reprojections such as in the example :ref:`sphx_glr_generated_gallery_map_transformations_reprojection_different_observers.py`.
However, this assumption is not appropriate for some observations, e.g., from coronagraphs.

There is now a context manager (:meth:`~sunpy.coordinates.frames.Helioprojective.assume_spherical_screen`) to override the default assumption such that any 2D coordinates are interpreted as being on the inside of a large spherical screen.
See the following example for how this context manager enables alternative reprojections.

.. minigallery:: sunpy.coordinates.Helioprojective.assume_spherical_screen

.. _whatsnew-2.1-performance:

Performance improvements
========================

> Several functions in `sunpy.map.maputils` have been significantly sped up with improved algorithms.

Reviewer: Erm...  `sunpy.map.maputils` isn't actually separately documented in the code reference.  Its functions are imported into `sunpy.map` and documented at that level.

> In addition, `sunpy.map.Map.wcs` is now cached when the map metadata remains unchanged, significantly improving performance in applications which make mutiple requests for the map WCS (e.g. plotting), and reducing the number of repeated warnings thrown when metadata is missing.

Reviewer: ```suggestion
In addition, `sunpy.map.GenericMap.wcs` is now cached when the map metadata remains unchanged, significantly improving performance in applications that make multiple requests for the map WCS (e.g., plotting), and reducing the number of repeated warnings thrown when metadata is missing.
```

.. _whatsnew-2.1-goes1617:

> Support for GOES-16 and GOES-17 X-Ray Sensor (XRS) data

Reviewer: This should be moved much much higher in the What's New, possibly even to be the first entry.

=======================================================
We have now included support for the GOES-16 and GOES-17 XRS 1-second data. This data can now be queried and downloaded with `sunpy.net.Fido` and the files read in and analysed as a `sunpy.timeseries.TimeSeries`.

> See the following example for how this can be achieved :ref:`sphx_glr_generated_gallery_acquiring_data_goes_xrs_example.py`

Reviewer: ```suggestion
See the following example for how this can be achieved: :ref:`sphx_glr_generated_gallery_acquiring_data_goes_xrs_example.py`
```

.. _whatsnew-2.1-python:

Increase in required package versions
=====================================
We have bumped the minimum version of several packages we depend on; these are the new minimum versions for sunpy 2.1:

- python 3.7
- astropy 4.0
- scipy 1.2
- parfive 1.1
- drms 0.6.1

> - matplotlib 3.1.1

Reviewer: Wait...  I thought we were bumping matplotlib with SunPy 3.0, not for SunPy 2.1...

Reviewer 2: ```suggestion
- matplotlib 2.2.2
```

.. doctest-skip-all

.. _whatsnew-2.1:

************************
What's New in SunPy 2.1?
************************

> 

Reviewer 3: ```suggestion
```

Overview
========
The SunPy project is pleased to announce the 2.1 release of the sunpy package.

On this page, you can read about some of the big changes in this release:

* :ref:`whatsnew-2.1-metadata`
* :ref:`whatsnew-2.1-refactor`
* :ref:`whatsnew-2.1-contours`
* :ref:`whatsnew-2.1-unitdata`
* :ref:`whatsnew-2.1-synoptic`
* :ref:`whatsnew-2.1-cutouts`
* :ref:`whatsnew-2.1-velocities`
* :ref:`whatsnew-2.1-reprojecting`
* :ref:`whatsnew-2.1-performance`
* :ref:`whatsnew-2.1-goes1617`
* :ref:`whatsnew-2.1-python`

SunPy 2.1 also includes a large number of smaller improvements and bug fixes, which are described in the :ref:`changelog`.

By the numbers:

* X commits have been added since 2.0
* X issues have been closed since 2.0
* X pull requests have been merged since 2.0
* X people have contributed since 2.0

> * X new contributors

Reviewer 3: Need to fill these in.

Performance improvements
========================
Several functions in `sunpy.map.maputils` have been significantly sped up with improved algorithms.

In addition, `sunpy.map.Map.wcs` is now cached when the map metadata remains unchanged, significantly improving performance in applications which make mutiple requests for the map WCS (e.g. plotting), and reducing the number of repeated warnings thrown when metadata is missing.

.. _whatsnew-2.1-goes1617:

Support for GOES-16 and GOES-17 X-Ray Sensor (XRS) data
=======================================================
We have now included support for the GOES-16 and GOES-17 XRS 1-second data. This data can now be queried and downloaded with `sunpy.net.Fido` and the files read in and analysed as a `sunpy.timeseries.TimeSeries`.

See the following example for how this can be achieved :ref:`sphx_glr_generated_gallery_acquiring_data_goes_xrs_example.py`

.. _whatsnew-2.1-python:

Increase in required package versions
=====================================
We have bumped the minimum version of several packages we depend on; these are the new minimum versions for sunpy 2.1:

- python 3.7
- astropy 4.0
- scipy 1.2
- parfive 1.1

> - drms 0.6

Reviewer 3: We should bump this to 0.6.1.

See the following example for how this context manager enables alternative reprojections.

.. minigallery:: sunpy.coordinates.Helioprojective.assume_spherical_screen

Performance improvements
========================
Several functions in `sunpy.map.maputils` have been significantly sped up with improved algorithms.

In addition, `sunpy.map.Map.wcs` is now cached when the map metadata remains unchanged, significantly improving performance in applications which make mutiple requests for the map WCS (e.g. plotting), and reducing the number of repeated warnings thrown when metadata is missing.

Support for GOES-16 and GOES-17 X-Ray Sensor (XRS) data
=======================================================
We have now included support for the GOES-16 and GOES-17 XRS 1-second data. This data can now be queried and downloaded with `sunpy.net.Fido` and the files read in and analysed as a `sunpy.timeseries.TimeSeries`.

See the following example for how this can be achieved:

> (add example "goes_xrs_example")

Reviewer 4: How can you link an example? (https://docs.sunpy.org/en/latest/generated/gallery/acquiring_data/goes_xrs_example.html#sphx-glr-generated-gallery-acquiring-data-goes-xrs-example-py)

Author: https://sphinx-gallery.github.io/stable/advanced.html#know-your-gallery-files

Reviewer 4: thanks! hmm still a bit confused though (sorry my sphynx knowledge is not great!) - is the best way just to do a 
```
.. literalinclude:: ../examples/acquiring_data/goes_xrs_example.py
   :language: python
```

or should I just hyperlink it?

Author: `literalinclude` will literally include (😆) the code or a code snippet, but I'm guessing you don't want to list all the code from the example here? If you just want a link you can use \`:ref:\`sphx_glr_gallery_acquiring_data_goes_xrs_example.py\` (which I think is the correctly parsed link).

Author: Yep, I've just checked an old locally generated version. If you build the docs and go to `docs/generated/gallery/acquiring_data/goes_xrs_example.rst`, the reference is given on line 9 as `sphx_glr_generated_gallery_acquiring_data_goes_xrs_example.py`

Reviewer: In case you need an example of how to type the reference, look at this line earlier in this file:
https://github.com/sunpy/sunpy/blame/842070404e970f6811463613feb2a73fc485961c/docs/whatsnew/2.1.rst#L65

Reviewer: I just edited @dstansby 's earlier comment to get around the fiddling with backticks
